test/unittests/capi_test.cpp

```diff
@@ -12,3 +16,224 @@ TEST(capi, validate)
+    EXPECT_NE(module, nullptr);
+
+    auto instance = fizzy_instantiate(module, nullptr, 0);
+    EXPECT_NE(instance, nullptr);
+
+    EXPECT_THAT(fizzy_execute(instance, 0, nullptr, 0, 0), Result());
+    EXPECT_THAT(fizzy_execute(instance, 1, nullptr, 0, 0), Result(42));
+    fizzy_value args[] = {{42}, {2}};
+    EXPECT_THAT(fizzy_execute(instance, 2, args, 2, 0), Result(21));
+    EXPECT_THAT(fizzy_execute(instance, 3, nullptr, 0, 0), Traps());
+
+    fizzy_free_instance(instance);
+}
+
+TEST(capi, execute_with_host_function)
+{
+    /* wat2wasm
+      (func (import "mod1" "foo1") (result i32))
+      (func (import "mod1" "foo2") (param i32 i32) (result i32))
+    */
+    const auto wasm = from_hex(
+        "0061736d01000000010b026000017f60027f7f017f021902046d6f643104666f6f310000046d6f643104666f6f"
+        "320001");
+    auto module = fizzy_parse(wasm.data(), wasm.size());
+    EXPECT_NE(module, nullptr);
+
+    fizzy_external_function host_funcs[] = {
+        {[](void*, fizzy_instance*, const fizzy_value*, uint32_t, int) {
+             return fizzy_execution_result{false, true, {42}};
+         },
+            nullptr},
+        {[](void*, fizzy_instance*, const fizzy_value* args, uint32_t, int) {
+             return fizzy_execution_result{false, true, {args[0].i64 / args[1].i64}};
+         },
+            nullptr}};
+
+    auto instance = fizzy_instantiate(module, host_funcs, 2);
+    EXPECT_NE(instance, nullptr);
+
+    EXPECT_THAT(fizzy_execute(instance, 0, nullptr, 0, 0), Result(42));
+
+    fizzy_value args[] = {{42}, {2}};
+    EXPECT_THAT(fizzy_execute(instance, 1, args, 2, 0), Result(21));
+
+    fizzy_free_instance(instance);
+}
+
+TEST(capi, imported_function_traps)
+{
+    /* wat2wasm
+      (func (import "m" "foo") (result i32))
+      (func (result i32)
+        call 0
+      )
+    */
+    const auto wasm =
+        from_hex("0061736d010000000105016000017f020901016d03666f6f0000030201000a0601040010000b");
+    auto module = fizzy_parse(wasm.data(), wasm.size());
+    EXPECT_NE(module, nullptr);
+
+    fizzy_external_function host_funcs[] = {
+        {[](void*, fizzy_instance*, const fizzy_value*, uint32_t, int) {
+             return fizzy_execution_result{true, false, {}};
+         },
+            nullptr}};
+
+    auto instance = fizzy_instantiate(module, host_funcs, 1);
+    EXPECT_NE(instance, nullptr);
+
+    EXPECT_THAT(fizzy_execute(instance, 1, nullptr, 0, 0), Traps());
+
+    fizzy_free_instance(instance);
+}
+
+TEST(capi, imported_function_void)
+{
+    /* wat2wasm
+      (func (import "m" "foo"))
+      (func
+        call 0
+      )
+    */
+    const auto wasm =
+        from_hex("0061736d01000000010401600000020901016d03666f6f0000030201000a0601040010000b");
+    auto module = fizzy_parse(wasm.data(), wasm.size());
+    EXPECT_NE(module, nullptr);
+
+    bool called = false;
+    fizzy_external_function host_funcs[] = {
+        {[](void* context, fizzy_instance*, const fizzy_value*, uint32_t, int) {
```

Comment:
What is this `context` doing?

Reply:
This is some opaque context from the host side that host can set during instantiate.

Here's an example where it's needed - calling a function from another instance:
https://github.com/wasmx/fizzy/blob/4cce6ee6a7a66b29cbc107808851454f30981c1e/test/unittests/capi_test.cpp#L226

(However, in this example called function's index is hardcoded, I expect it's going to be inconvenient in case you would need to pass both instance and function index via one `context` pointer - say if you want a single general host function dispatching the calls to another instance)